chpl/chpl-service/src/main/java/gov/healthit/chpl/directreview/DirectReviewListingSharedStoreHandler.java

```diff
@@ -0,0 +1,81 @@
+package gov.healthit.chpl.directreview;
+
+import java.util.Date;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import gov.healthit.chpl.certifiedproduct.CertifiedProductDetailsManager;
+import gov.healthit.chpl.domain.CertifiedProductSearchDetails;
+import gov.healthit.chpl.domain.compliance.DirectReview;
+import gov.healthit.chpl.exception.EntityRetrievalException;
+import gov.healthit.chpl.sharedstore.listing.SharedListingStoreProvider;
+import gov.healthit.chpl.util.NullSafeEvaluator;
+import lombok.extern.log4j.Log4j2;
+
+@Component
+@Log4j2
+public class DirectReviewListingSharedStoreHandler {
+    private SharedListingStoreProvider sharedListingStoreProvider;
+    private CertifiedProductDetailsManager certifiedProductDetailsManager;
+
+    @Autowired
+    public DirectReviewListingSharedStoreHandler(@Lazy SharedListingStoreProvider sharedListingStoreProvider,
+            @Lazy CertifiedProductDetailsManager certifiedProductDetailsManager) {
+        this.sharedListingStoreProvider = sharedListingStoreProvider;
+        this.certifiedProductDetailsManager = certifiedProductDetailsManager;
+    }
+
+    public void handler(List<DirectReview> allDirectReviews) {
+        getUniqueListings(allDirectReviews).forEach(listing -> {
+            removeListingFromSharedStoreIfDirectReviewUpdated(listing, allDirectReviews);
+        });
+    }
+
+    private void removeListingFromSharedStoreIfDirectReviewUpdated(CertifiedProductSearchDetails listing, List<DirectReview> allDirectReviews) {
+        listing.getDirectReviews().forEach(dr -> {
+            if (hasDirectReviewBeenUpdated(findDirectReview(allDirectReviews, dr.getJiraKey()), dr)) {
+                LOGGER.info("Removing Listing Id {} from the Shared Store", listing.getId());
+                sharedListingStoreProvider.remove(listing.getId());
+            } else {
+                LOGGER.info("Not Removing Listing Id {} from the Shared Store - Direct Review Last Updated Date not changed", listing.getId());
+            }
+        });
+    }
+
+    private Boolean hasDirectReviewBeenUpdated(DirectReview newVersion, DirectReview origVersion) {
+        return ! NullSafeEvaluator.eval(() -> newVersion.getLastUpdated(), new Date(Long.MIN_VALUE)).equals(
+                NullSafeEvaluator.eval(() -> origVersion.getLastUpdated(), new Date(Long.MIN_VALUE)));
+
+    }
+
+    private DirectReview findDirectReview(List<DirectReview> directReviews, String jiraKey) {
+        return directReviews.stream()
+                .filter(dr -> dr.getJiraKey().equals(jiraKey))
+                .findAny()
+                .orElse(null);
+    }
+
+    private List<CertifiedProductSearchDetails> getUniqueListings(List<DirectReview> allDirectReviews) {
```

Comment:
I think this method is trying to get all the listings related to all direct reviews to potentially remove those listings from the shared store. There is a little bit more complicated logic to say "is a direct review associated to this listing" than looking at the Developer-Associated Listings. The logic for that is encapsulated in the DirectReviewSearchService getDirectReviewsRelatedToListing method (I think the method's comment block is accurate).  I wonder if you could make use of that method in some way? If a Direct Review doesn't have any Developer-Associated Listings it can still be associated to listings related to the developer and I think we would want to clear those listings from the shared store.